core-plugins/src/main/java/co/cask/hydrator/plugin/batch/joiner/Joiner.java

```diff
@@ -0,0 +1,289 @@
+      Iterable<String> multipleJoinKeys = Splitter.on('&').trimResults().omitEmptyStrings().split(conf.getJoinKeys());
+      if (Iterables.isEmpty(multipleJoinKeys)) {
+        throw new IllegalArgumentException("Join keys can not be empty.");
+      }
+      int numJoinKeys = 0;
+      for (String singleJoinKey : multipleJoinKeys) {
+        KeyValueListParser kvParser = new KeyValueListParser("\\s*=\\s*", "\\.");
+        Iterable<KeyValue<String, String>> keyValues = kvParser.parse(singleJoinKey);
+        if (numJoinKeys == 0) {
+          numJoinKeys = Iterables.size(keyValues);
+        } else if (numJoinKeys != Iterables.size(keyValues)) {
+          throw new IllegalArgumentException("There should be one join key from each of the stages. Please add join " +
+                                               "keys for each stage.");
+        }
+        for (KeyValue<String, String> keyValue : keyValues) {
+          String stageName = keyValue.getKey();
+          String joinKey = keyValue.getValue();
+          if (!stageToKey.containsKey(stageName)) {
+            stageToKey.put(stageName, new ArrayList<String>());
+          }
+          stageToKey.get(stageName).add(joinKey);
+        }
+      }
+    } else {
+      throw new IllegalArgumentException("Join keys can not be empty");
+    }
+    return stageToKey;
+  }
+
+  private Table<String, String, String> getPerStageFieldsToRename() {
+    // table to store <stageName, oldFieldName, alias>
+    Table<String, String, String> perStageFieldsToRename = HashBasedTable.create();
+
+    if (!Strings.isNullOrEmpty(conf.getSelectedFields())) {
+      Map<String, Map<String, String>> stageToRenameFields = GSON.fromJson(conf.getSelectedFields(), MAP_STRING_MAP);
+      for (Map.Entry<String, Map<String, String>> stageToRenameField : stageToRenameFields.entrySet()) {
+        Map<String, String> oldToAliases = stageToRenameField.getValue();
+        for (Map.Entry<String, String> entry : oldToAliases.entrySet()) {
+          if (Strings.isNullOrEmpty(entry.getKey()) || Strings.isNullOrEmpty(entry.getValue())) {
+            throw new IllegalArgumentException("Either field name or alias is missing.");
+          }
+          perStageFieldsToRename.put(stageToRenameField.getKey(), entry.getKey(), entry.getValue());
+        }
+      }
+    } else {
+      throw new IllegalArgumentException("selectedFields can not be empty. Please provide at least 1 selectedFields");
+    }
+    return perStageFieldsToRename;
+  }
+
+  Iterable<String> getRequiredInputs() {
+    if (conf.getRequiredInputs() != null) {
+      return Splitter.on(',').trimResults().omitEmptyStrings().split(conf.getRequiredInputs());
+    }
+    return Collections.<String>emptyList();
+  }
+
+  @Path("outputSchema")
+  public Schema getOutputSchema(GetSchemaRequest request) {
+    try {
+      return getOutputSchema(request.inputSchemas);
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(e.getMessage());
```

Comment:
I think you should also send the `e` into the BadRequestException. can you do that? That way, the root stacktrace isn't lost